### PR TITLE
Don't include snippets of source in diagnostics about types

### LINKED ISSUE #7387: C# verbatim character should be removed in error messages
Consider the following:
```
[@X] class F {}
```
Roslyn reports that `@X` cannot be found, whereas the legacy compiler reports that `X` cannot be found. 

Where spotted:
- CS0246
- CS0579
<strike>
- CS0164
</strike>
- CA9999

### PR DESCRIPTION
**Pull request by gafter**
Fixes #7387 

@dotnet/roslyn-compiler Please review.

### CODE REVIEW

File: src/Compilers/CSharp/Portable/Syntax/GenericNameSyntax.cs [hunk 1]
```diff
@@ -15,5 +16,12 @@ public bool IsUnboundGenericName
                 return this.TypeArgumentList.Arguments.Any(SyntaxKind.OmittedTypeArgument);
             }
         }
+
+        internal override string ErrorDisplayName()
+        {
+            var pb = PooledStringBuilder.GetInstance();
+            pb.Builder.Append(Identifier.ValueText).Append("<").Append(new string(',', Arity - 1)).Append(">");
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Syntax/GenericNameSyntax.cs]: `StringBuilder` has `Append(char, int)` overload.

File: src/Compilers/CSharp/Portable/Syntax/IdentifierNameSyntax.cs [hunk 2]
```diff
@@ -0,0 +1,16 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using Microsoft.CodeAnalysis.CSharp.Symbols;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Text;
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Syntax/IdentifierNameSyntax.cs]: Usings are not needed.

File: src/Compilers/CSharp/Portable/Syntax/QualifiedNameSyntax.cs [hunk 3]
```diff
@@ -16,5 +16,10 @@ internal override SimpleNameSyntax GetUnqualifiedName()
         {
             return Right;
         }
+
+        internal override string ErrorDisplayName()
+        {
+            return Left.ErrorDisplayName() + "." + Right.ErrorDisplayName();
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Syntax/QualifiedNameSyntax.cs]: Are we testing this case and the `AliasedQualifiedNameSyntax` case?

**Comment by cston** [comment]
LGTM

File: src/Compilers/CSharp/Portable/Syntax/GenericNameSyntax.cs [hunk 4]
```diff
@@ -15,5 +16,12 @@ public bool IsUnboundGenericName
...
+            return pb.ToStringAndFree();
+        }
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Syntax/GenericNameSyntax.cs]: Are we testing these cases?

**Comment by leppie** [comment]
This changes allows for the following to compile incorrectly
```
class C { public static void Main() { @var v = 1; } }
```
This is slightly modified version of the failing code action test case.

I think the error should still be CS0246.

**Comment by leppie** [comment]
This is failing due to an incorrect check now for `var`. A simple fix is just to change (in `Binder.NotFound`):
```
if (whereText == "var" && !options.IsAttributeTypeLookup())
```
to
```
if (where.ToString() == "var" && !options.IsAttributeTypeLookup())
```

Also consider adding the following (and more if there are) as a test:
```
class C { public static void Main() { @var v = 1; } }
```
to report CS0246

Note: Not sure how expensive `where.ToString()` is, but it was previously called. This fix would also explain why `whereText` was probably introduced in `Binder.NotFound`. Perhaps knowing the reason now, and that both callsites of this method passes the same value as `simpleName`, it can be removed.

**Comment by leppie** [comment]
This PR also seems to partially fix #7493 to bring it inline with what the legacy compiler reported.

File: src/Compilers/CSharp/Portable/Binder/Binder_Symbols.cs [hunk 5]
```diff
@@ -1524,7 +1524,7 @@ public int Compare(Symbol fst, Symbol snd)
                     node = node.Parent;
                 }
 
-                CSDiagnosticInfo info = NotFound(where, simpleName, arity, where.ToString(), diagnostics, aliasOpt, qualifierOpt, options);
+                CSDiagnosticInfo info = NotFound(where, simpleName, arity, simpleName, diagnostics, aliasOpt, qualifierOpt, options);
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/Binder_Symbols.cs]: It looks like there are only two call sites for this function and now both are passing the same value for the second and the forth parameter. We should probably get rid of one of them to simplify things.

File: src/Compilers/CSharp/Portable/Binder/Binder_Symbols.cs [hunk 6]
```diff
@@ -1524,7 +1524,7 @@ public int Compare(Symbol fst, Symbol snd)
...
                 return new ExtendedErrorTypeSymbol(qualifierOpt ?? Compilation.Assembly.GlobalNamespace, simpleName, arity, info);
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/Binder_Symbols.cs]: It looks like lines 1285, 1424, 1431 are also using syntax node as a diagnostics argument. Should they be changed too?

File: src/Compilers/CSharp/Portable/Syntax/AttributeSyntax.cs [hunk 7]
```diff
@@ -18,7 +15,7 @@ public partial class AttributeSyntax
         internal string GetErrorDisplayName()
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Syntax/AttributeSyntax.cs]: Should this be simply an override of ErrorDisplayName?

Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Syntax/AttributeSyntax.cs]: I guess not, because this type doesn't derive from NameSyntax.

File: src/Compilers/CSharp/Test/Symbol/Symbols/Metadata/PE/TypeForwarders.cs [hunk 8]
```diff
@@ -795,25 +795,26 @@ class Test
...
                 //     Outer.Inner<string> F;
-                Diagnostic(ErrorCode.WRN_UnreferencedField, "F").WithArguments("Test.F"),
+                Diagnostic(ErrorCode.WRN_UnreferencedField, "F").WithArguments("Test.F").WithLocation(6, 25),
                 // (8,5): error CS0246: The type or namespace name 'Generic' could not be found (are you missing a using directive or an assembly reference?)
                 //     Generic G0 { get; set; }
-                Diagnostic(ErrorCode.ERR_SingleTypeNameNotFound, "Generic").WithArguments("Generic"),
-                // (9,5): error CS1070: The type name 'Generic<int>' could not be found. This type has been forwarded to assembly 'pe2, Version=0.0.0.0, Culture=neutral, PublicKeyToken=null'. Consider adding a reference to that assembly.
+                Diagnostic(ErrorCode.ERR_SingleTypeNameNotFound, "Generic").WithArguments("Generic").WithLocation(8, 5),
+                // (9,5): error CS1070: The type name 'Generic' could not be found. This type has been forwarded to assembly 'pe2, Version=0.0.0.0, Culture=neutral, PublicKeyToken=null'. Consider adding a reference to that assembly.
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Symbol/Symbols/Metadata/PE/TypeForwarders.cs]: It feels like in this case it would be better to say 'Generic<>' is not found instead of using simple name.

File: src/Compilers/CSharp/Test/Symbol/Symbols/GenericConstraintTests.cs [hunk 9]
```diff
@@ -2522,14 +2522,19 @@ class C<T>
 }";
             CreateCompilationWithMscorlib(source).VerifyDiagnostics(
                 // (4,15): error CS0246: The type or namespace name 'X' could not be found (are you missing a using directive or an assembly reference?)
+                //     where T : X
                 Diagnostic(ErrorCode.ERR_SingleTypeNameNotFound, "X").WithArguments("X").WithLocation(4, 15),
-                // (5,15): error CS0246: The type or namespace name 'I<T>' could not be found (are you missing a using directive or an assembly reference?)
-                Diagnostic(ErrorCode.ERR_SingleTypeNameNotFound, "I<T>").WithArguments("I<T>").WithLocation(5, 15),
+                // (5,15): error CS0246: The type or namespace name 'I' could not be found (are you missing a using directive or an assembly reference?)
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Symbol/Symbols/GenericConstraintTests.cs]: It feels like in this case it would be better to say 'I<>' is not found instead of using simple name.

Comment (leppie) [inline src/Compilers/CSharp/Test/Symbol/Symbols/GenericConstraintTests.cs]: @AlekseyTs There is an issue #7493 for suggestions. `I<>` could work, but this might be confusing.

File: src/Compilers/CSharp/Portable/Syntax/GenericNameSyntax.cs (hunk 4 above)
Comment (gafter) [inline src/Compilers/CSharp/Portable/Syntax/GenericNameSyntax.cs]: Yes, some of the changed tests are due to this.

File: src/Compilers/CSharp/Portable/Syntax/GenericNameSyntax.cs (hunk 1 above)
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Portable/Syntax/GenericNameSyntax.cs]: Why is this the error display name?  It seems appropriate for an unbound generic. However, if this is a bound generic, why not have the type argument list?

Comment (gafter) [inline src/Compilers/CSharp/Portable/Syntax/GenericNameSyntax.cs]: Because when we are unable to *resolve* the type name, the actual type arguments are irrelevant.

File: src/Compilers/CSharp/Portable/Syntax/QualifiedNameSyntax.cs (hunk 3 above)
Comment (gafter) [inline src/Compilers/CSharp/Portable/Syntax/QualifiedNameSyntax.cs]: Some of the test changes are due to this.

File: src/Compilers/CSharp/Portable/Binder/Binder_Symbols.cs (hunk 5 above)
Comment (gafter) [inline src/Compilers/CSharp/Portable/Binder/Binder_Symbols.cs]: Not so. One of the call sites looks like this:

```cs
                NotFound(where, simpleName, arity, whereText + "Attribute", diagnostics, aliasOpt, qualifierOpt, options | LookupOptions.VerbatimNameAttributeTypeOnly);
```

File: src/Compilers/CSharp/Portable/Binder/Binder_Symbols.cs (hunk 6 above)
Comment (gafter) [inline src/Compilers/CSharp/Portable/Binder/Binder_Symbols.cs]: There are no compiler error codes between 1113 and 1501, so I'm not sure what you're referring to.  From ErrorCode.cs:

```cs
        ERR_ExplicitExtension = 1112,
        ERR_ValueTypeExtDelegate = 1113,
        // unused 1114-1199
        // Below five error codes are unused.
        // WRN_FeatureDeprecated2 = 1200,
        // WRN_FeatureDeprecated3 = 1201,
        // WRN_FeatureDeprecated4 = 1202,
        // WRN_FeatureDeprecated5 = 1203,
        // WRN_OldWarning_FeatureDefaultDeprecated = 1204,
        // unused 1205-1500
        ERR_BadArgCount = 1501,
        //ERR_BadArgTypes = 1502,
        ERR_BadArgType = 1503,
        ERR_NoSourceFile = 1504,
```

**Comment by gafter** [comment]
@dotnet/roslyn-compiler Could I have a second review, please?

File: src/Compilers/CSharp/Portable/Binder/Binder_Symbols.cs (hunk 6 above)
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/Binder_Symbols.cs]: I am referring to lines of code in this file.

File: src/Compilers/CSharp/Portable/Binder/Binder_Symbols.cs [hunk 10]
```diff
@@ -1524,7 +1524,7 @@ public int Compare(Symbol fst, Symbol snd)
                     node = node.Parent;
                 }
 
-                CSDiagnosticInfo info = NotFound(where, simpleName, arity, where.ToString(), diagnostics, aliasOpt, qualifierOpt, options);
+                CSDiagnosticInfo info = NotFound(where, simpleName, arity, (where as NameSyntax)?.ErrorDisplayName() ?? simpleName, diagnostics, aliasOpt, qualifierOpt, options);
                 return new ExtendedErrorTypeSymbol(qualifierOpt ?? Compilation.Assembly.GlobalNamespace, simpleName, arity, info);
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/Binder_Symbols.cs]: It looks like **lines** of code 1285, 1424, 1431 are also using syntax node as a diagnostics argument. Should they be changed too?

File: src/Compilers/CSharp/Test/Semantic/Semantics/LookupTests.cs [hunk 11]
```diff
@@ -1424,6 +1424,16 @@ static int Main(string[] args)
             Assert.Equal(CandidateReason.OverloadResolutionFailure, symbolInfo.CandidateReason);
         }
 
+        public void TestLookupVerbatimVar()
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/Semantics/LookupTests.cs]: This method is not properly marked as a test method.

File: src/Compilers/CSharp/Test/Syntax/Parsing/ParserRegressionTests.cs [hunk 12]
```diff
@@ -80,7 +80,7 @@ public void c024928()
         }
 
         [WorkItem(2771, "https://github.com/dotnet/roslyn/issues/2771")]
-        [Fact]
+        [Fact(Skip = "This test continues to be flaky, even in this form, due to variations in the amount of memory available when it is run")]
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Syntax/Parsing/ParserRegressionTests.cs]: Please excluding this change from this commit. If you wish to disable the test, please do this in a separate commit referencing the issue that tracks the flakiness of the test (the issue probably already exists because @TyOverby is working on addressing the flakiness).

**Comment by gafter** [comment]
@dotnet-bot test this please

File: src/Compilers/CSharp/Portable/Binder/Binder_Symbols.cs (hunk 10 above)
Comment (gafter) [inline src/Compilers/CSharp/Portable/Binder/Binder_Symbols.cs]: @AlekseyTs Those are fixed now, too. Can you please re-review?

File: src/Compilers/CSharp/Test/Semantic/Semantics/LookupTests.cs [hunk 13]
```diff
@@ -1424,6 +1424,17 @@ static int Main(string[] args)
             Assert.Equal(CandidateReason.OverloadResolutionFailure, symbolInfo.CandidateReason);
         }
 
+        [Fact]
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/Semantics/LookupTests.cs]: Should this test be linked to the issue?

**Comment by AlekseyTs** [comment]
It looks like the baseline for Microsoft.CodeAnalysis.CSharp.UnitTests.CompilationErrorTests.CS0104ERR_AmbigContext01 test should be adjusted.

**Comment by AlekseyTs** [comment]
LGTM, modulo the base line adjustment for CS0104ERR_AmbigContext01.
